File: C#/Recursive/Collatz_rec.cs
//Author: Jarrett Melnick
//Program to return the top 10 numbers with the largest Collatz sequence in a given range
using System;

public class Collatz{
 	public static void Main(){
		Console.Write("Enter the maximum value to be considered: ");
		long iterate = Convert.ToInt64(Console.ReadLine());	

		number[] result = new number[10]; 

		for(int i = 0; i < 10; i++){
			result[i] = new number();
		}

		for(long i = 2; i <= iterate; i++){
			long curCollatz = CollatzConjecture(i,0);
			if((curCollatz > result[9].collatzVal) && !(hasCollatzNumber(result, curCollatz))){
				number num = new number(i, curCollatz);
				result[9] = num;
				result = sorter(result);
			}
		}
		//Prints the top 10 numbers with the largest Collatz sequences.
		result = sorter(result);
		for(int i = 0; i < 10; i++){
			Console.WriteLine(i+1 + " is " + result[i].identity + 
				" and has a collatz conjecture of " + result[i].collatzVal);
		}
 	}
 	public static long CollatzConjecture(long val, long collatz){
 		//Function to calculate the length of a given number's collatz sequence.
 		string binNumber = Convert.ToString(val, 2);
		if(val <= 1){
			return collatz;
		}
		else{
			return (binNumber.Trim().EndsWith("1")) ? (CollatzConjecture(3*val+1,collatz+1)): (CollatzConjecture(val/2,collatz+1));
		}
 	}
 	 	private static bool hasCollatzNumber(number[] num, long val){
		//Determines if the given array num has a number with a Collatz sequence of length val.
 		for(int i = 0; i < 10; i++){
 			if(num[i].collatzVal == val)
 			return true;
 		}
 		return false;
 	}
 	private static number[] sorter(number[] num){
 		//Sorts the array by the number with the largest Collatz sequence
		bool sorted = false;
		while(!sorted){
			sorted = true;

			for(int i = 1; i < 10 ; i++){
				if(num[i].collatzVal > num[i-1].collatzVal){
					number temp = num[i-1];
					num[i-1] = num[i];
					num[i] = temp;
					sorted = false;
				}
			}
		}
		return num;
	}

}
class number{
	//Class to hold data for Collatz sequence
	public long identity = 0;		//Identity of number
	public long collatzVal = 0;		//Length of Collatz value
	public number(){
		identity = 0;
		collatzVal = 0;
	}
	public number(long ident, long colVal){
		//Full constructor with provided identity and collatz value.
		identity = ident;
		collatzVal = colVal;
	}
	public long getColatzVal(){
		//Returns collatz value
		return collatzVal;
	}
}

File: Julia/Recursive/Collatz_rec.jl
#Author: Jarrett Melnick
#Program to list the top 10 numbers with the largest Collatz sequences in a given range.

#Function to perform a Collatz sequence on a number and calculate the length of the sequence
collatz(n) = (n > 2) ? ((n%2) == 1? (collatz(3*n+1) + 1): (collatz(n >> 1) + 1)): n-1

function smallest(numArray)
    #Returns the index of the smallest value in array numArray
    mini = 1
    for i = 1:10
        if(numArray[i] < numArray[mini])
            mini = i
        end
    end
    return mini
end
function largest(numArray)
    #Returns the index of the largest value in array numArray
    maxi = 1
    for i = 1:10
        if(numArray[i] > numArray[maxi])
            maxi = i
        end
    end
    return maxi
end
function hasSequenceLen(numArray, val)
    #Returns true if the array numArray has the number val
    hasVal = false
    for i = 1:10
        if(numArray[i] == val)
            hasVal = true
        end
    end
    return hasVal
end
function printTable(numArrayI, numArrayS)
    for i = 1:10
        @printf "%d has a length of %d\n" numArrayI[i] numArrayS[i] 
    end
end

#Program logic:
print("Please input the max value to be considered: ")
iterate = readline()
iterate = parse(Int64, chomp(iterate))
identities = [0,0,0,0,0,0,0,0,0,0]
sequences = [0,0,0,0,0,0,0,0,0,0]

println("Calculating Sequences...")
for i = 2:iterate
    curCollatz = collatz(i)
    small = smallest(sequences)
    hasSequence = hasSequenceLen(sequences, curCollatz)
    if((curCollatz > sequences[small]) && !(hasSequence))
        sequences[small] = curCollatz
        identities[small] = i
    end
end
println("Printing Table...")
for i = 1:10
    large = largest(sequences)
    @printf "%d has a Collatz sequence of length %d\n" identities[large] sequences[large]
    identities[large] = 0
    sequences[large] = 0
end